Stop using GraphArgs for shape env guard source tracking

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90913
* #90912
* __->__ #90911

GraphArgs worked fairly well, but it was still missing sources
sometimes.  Now, we maintain an auxiliary data structure which we
MUST populate whenever we fakeify a tensor / allocate a bare SymInt.
This should guarantee once and for all that every symbol is available.
Should fix swin_base_patch4_window7_224.

While I was at it, I moved fakeification utility back to builder
as it was only used at once call site.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

ezyang committed:
Stop using GraphArgs for shape env guard source tracking

GraphArgs worked fairly well, but it was still missing sources
sometimes.  Now, we maintain an auxiliary data structure which we
MUST populate whenever we fakeify a tensor / allocate a bare SymInt.
This should guarantee once and for all that every symbol is available.
Should fix swin_base_patch4_window7_224.

While I was at it, I moved fakeification utility back to builder
as it was only used at once call site.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Stop using GraphArgs for shape env guard source tracking"


GraphArgs worked fairly well, but it was still missing sources
sometimes.  Now, we maintain an auxiliary data structure which we
MUST populate whenever we fakeify a tensor / allocate a bare SymInt.
This should guarantee once and for all that every symbol is available.
Should fix swin_base_patch4_window7_224.

While I was at it, I moved fakeification utility back to builder
as it was only used at once call site.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on test/dynamo/test_repros.py:
```diff
@@ -2030,6 +2030,22 @@ def f(x, y):
         y = torch.randn(2)
         self.assertEqual(f(x, y), opt_f(x, y))
 
+    def test_swin_base_tensor_attr(self):
```
swin?

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -920,3 +925,36 @@ def _clone_input(value):
+# Tracks the sources of all fake tensors we wrap in Dynamo.
+# Used by shape guard computation.
+@dataclasses.dataclass
+class FakeSource:
+    fake: Union[FakeTensor, SymInt]
+    sname: str
```
nit: maybe put it in source.py?

ezyang replied:
it's mostly here because GraphArgs is here, I don't have a big opinion (besides avoiding cycles)

ezyang replied:
Hmm, actually there's some undesirable crosstalk here; this is not a subclass of Source. I'll give it another name.

voznesenskym commented on torch/_dynamo/variables/builder.py:
```diff
@@ -920,3 +925,36 @@ def _clone_input(value):
+    sname: str
+
+
+def wrap_to_fake_tensor_and_record(
```
If this were my PR, you would get me for unnecessary code motion :P

ezyang replied:
:P

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -920,3 +925,36 @@ def _clone_input(value):
+    sname: str
+
+
+def wrap_to_fake_tensor_and_record(
```
yes

ezyang committed:
Update on "Stop using GraphArgs for shape env guard source tracking"


GraphArgs worked fairly well, but it was still missing sources
sometimes.  Now, we maintain an auxiliary data structure which we
MUST populate whenever we fakeify a tensor / allocate a bare SymInt.
This should guarantee once and for all that every symbol is available.
Should fix swin_base_patch4_window7_224.

While I was at it, I moved fakeification utility back to builder
as it was only used at once call site.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
